### PR TITLE
[android] conditionally build DS2 for the Android SDK

### PR DESCRIPTION
**Pull request by andrurogerz**
Conditionally include the [DS2debug server](https://github.com/compnerd/ds2) in the Android Swift SDK build. DS2 and its dependencies are only included when `build.ps1` is invoked with the `-IncludeDS2` flag.

This change is comprised of three primary parts:
1. Fetch and unpack WinFlexBison tools; `flex` and `bison` equivalents are required to build DS2.
2. Build the `RegsGen2` host tool, specifying the previously unpacked `win_flex.exe` and `win_bison.exe` tools.
3. Build ds2 binary for each each requested Android SDK architecture and install it to the proper location under the `Android.platform` directory.

NOTE: the Android SDK isn't currently included in the installer build, so the ds2 binary only ends up in the output directory and not in any .msi file.

## Test Plan
Build with and without ds2 included
```
cd S:\SourceCache
.\swift\utils\build.cmd -AndroidSDKs aarch64,x86_64
.\swift\utils\build.cmd -AndroidSDKs aarch64,x86_64 -IncludeDS2
```
Confirm the ds2 binary appears in the expected location only when `-IncludeDS2` is specified
```
cd S:\b
dir /s /b ds2
S:\b\aarch64\Android\ds2
S:\b\aarch64\Android\ds2\ds2
S:\b\arm64\Android.platform\Developer\Library\aarch64-unknown-linux-android\bin\ds2
S:\b\x64\Android.platform\Developer\Library\x86_64-unknown-linux-android\bin\ds2
S:\b\x86_64\Android\ds2
S:\b\x86_64\Android\ds2\ds2
```

### CODE REVIEW

**Comment by andrurogerz** [comment]
@hjyamauchi would you mind having a look at this change?

@shahmishal I believe you have some stake in `build.ps1` changes not breaking CI. This change is meant to have no effect unless the new argument is provided, so should be a noop for you.

File: utils/build.ps1 [hunk 1]
```diff
@@ -1490,6 +1515,39 @@ function Build-XML2([Platform]$Platform, $Arch) {
...
+  Build-CMakeProject `
+    -Src "$SourceCache\ds2" `
+    -Bin "$($Arch.BinaryCache)\$Platform\ds2" `
+    -InstallTo "$($Arch.PlatformInstallRoot)\Developer\Library\$ArchName" `
+    -Arch $Arch `
+    -Platform $Platform `
+    -BuildTargets default `
+    -Defines @{
+      CMAKE_SYSTEM_NAME = $Platform.ToString();
+      DS2_REGSGEN2 = "$(Get-HostProjectBinaryCache RegsGen2)/regsgen2.exe"
```
Comment (hjyamauchi) [inline utils/build.ps1:1545]: Could have a `;` at the end of the line?

**Review by hjyamauchi — APPROVED** [review]

**Comment by hjyamauchi** [comment]
LGTM. Note the `IncludeDS2` flag is an opt-in (off by default.)

**Comment by hjyamauchi** [comment]
@swift-ci please test

**Comment by andrurogerz** [comment]
@swift-ci please test

File: utils/build.ps1 [hunk 2]
```diff
@@ -335,6 +340,16 @@ function Get-AndroidNDKPath {
   return $androidNDKPath
 }
 
+function Get-FlexExecutable {
+  $flexExecutable = Join-Path -Path $BinaryCache -ChildPath "win_flex_bison\win_flex.exe"
+  return $flexExecutable
```
Comment (compnerd) [inline utils/build.ps1:345]:
```suggestion
  return Join-Path -Path $BinaryCache -ChildPath "win_flex_bison\win_flex.exe"
```

File: utils/build.ps1 [hunk 3]
```diff
@@ -335,6 +340,16 @@ function Get-AndroidNDKPath {
...
+}
+
+function Get-BisonExecutable {
+  $bisonExecutable = Join-Path -Path $BinaryCache -ChildPath "win_flex_bison\win_bison.exe"
+  return $bisonExecutable
```
Comment (compnerd) [inline utils/build.ps1:350]:
```suggestion
  return Join-Path -Path $BinaryCache -ChildPath "win_flex_bison\win_bison.exe"
```

File: utils/build.ps1 [hunk 4]
```diff
@@ -1490,6 +1515,39 @@ function Build-XML2([Platform]$Platform, $Arch) {
...
+  Build-CMakeProject `
+    -Src "$SourceCache\ds2" `
+    -Bin "$($Arch.BinaryCache)\$Platform\ds2" `
+    -InstallTo "$($Arch.PlatformInstallRoot)\Developer\Library\$ArchName" `
+    -Arch $Arch `
+    -Platform $Platform `
+    -BuildTargets default `
+    -Defines @{
+      CMAKE_SYSTEM_NAME = $Platform.ToString();
+      DS2_REGSGEN2 = "$(Get-HostProjectBinaryCache RegsGen2)/regsgen2.exe";
```
Comment (compnerd) [inline utils/build.ps1:1545]: This seems wrong - this should be using `Get-BuildProjectBinaryCache` as the tool is being run on build, not host.

File: utils/build.ps1 [hunk 5]
```diff
@@ -2471,6 +2529,10 @@ if ($Clean) {
   }
 }
 
+if (-not $SkipBuild -and $IncludeDS2) {
+  Invoke-BuildStep Build-RegsGen2 $HostArch
+}
```
Comment (compnerd) [inline utils/build.ps1:2534]: Why not hoist this into the tools build above? I think it makes sense to build this around the compilers phase and in the same block.

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Review by compnerd — APPROVED** [review]

**Pull request merged**